internal/rbd/qos.go

```diff
@@ -0,0 +1,281 @@
+
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	librbd "github.com/ceph/go-ceph/rbd"
+)
+
+const (
+	// Qos parameters name of StorageClass.
+	baseReadIops            = "BaseReadIops"
+	baseWriteIops           = "BaseWriteIops"
+	baseReadBytesPerSecond  = "BaseReadBytesPerSecond"
+	baseWriteBytesPerSecond = "BaseWriteBytesPerSecond"
+	readIopsPerGB           = "ReadIopsPerGB"
+	writeIopsPerGB          = "WriteIopsPerGB"
+	readBpsPerGB            = "ReadBpsPerGB"
+	writeBpsPerGB           = "WriteBpsPerGB"
+	baseVolSizeBytes        = "BaseVolSizeBytes"
+
+	// Qos type name of rbd image.
+	readIopsLimit      = "rbd_qos_read_iops_limit"
+	writeIopsLimit     = "rbd_qos_write_iops_limit"
+	readBpsLimit       = "rbd_qos_read_bps_limit"
+	writeBpsLimit      = "rbd_qos_write_bps_limit"
+	metadataConfPrefix = "conf_"
+
+	// The params use to calc qos based on capacity.
+	baseQosReadIopsLimit  = "rbd_base_qos_read_iops_limit"
+	baseQosWriteIopsLimit = "rbd_base_qos_write_iops_limit"
+	baseQosReadBpsLimit   = "rbd_base_qos_read_bps_limit"
+	baseQosWriteBpsLimit  = "rbd_base_qos_write_bps_limit"
+	readIopsPerGBLimit    = "rbd_read_iops_per_gb_limit"
+	writeIopsPerGBLimit   = "rbd_write_iops_per_gb_limit"
+	readBpsPerGBLimit     = "rbd_read_bps_per_gb_limit"
+	writeBpsPerGBLimit    = "rbd_write_bps_per_gb_limit"
+	baseQosVolSize        = "rbd_base_qos_vol_size"
+)
+
+type qosSpec struct {
+	baseLimitType  string
+	baseLimit      string
+	perGBLimitType string
+	perGBLimit     string
+	provide        bool
+}
+
+func parseQosParams(
+	scParams map[string]string,
+) map[string]*qosSpec {
+	rbdQosParameters := map[string]*qosSpec{
+		baseReadIops:            {readIopsLimit, "", readIopsPerGB, "", false},
+		baseWriteIops:           {writeIopsLimit, "", writeIopsPerGB, "", false},
+		baseReadBytesPerSecond:  {readBpsLimit, "", readBpsPerGB, "", false},
+		baseWriteBytesPerSecond: {writeBpsLimit, "", writeBpsPerGB, "", false},
+	}
+	for k, v := range scParams {
+		if qos, ok := rbdQosParameters[k]; ok && v != "" {
+			qos.baseLimit = v
+			qos.provide = true
+			for _k, _v := range scParams {
+				if _k == qos.perGBLimitType && _v != "" {
+					qos.perGBLimit = _v
+				}
+			}
+		}
+	}
+
+	return rbdQosParameters
+}
+
+func (rv *rbdVolume) SetQOS(
+	ctx context.Context,
+	scParams map[string]string,
+) error {
+	rv.BaseVolSize = ""
+	if v, ok := scParams[baseVolSizeBytes]; ok && v != "" {
+		rv.BaseVolSize = v
+	}
+
+	rbdQosParameters := parseQosParams(scParams)
+	for _, qos := range rbdQosParameters {
+		if qos.provide {
+			err := calcQosBasedOnCapacity(ctx, rv, *qos)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (rv *rbdVolume) ApplyQOS(
+	ctx context.Context,
+) error {
+	for k, v := range rv.QosParameters {
+		err := rv.SetMetadata(metadataConfPrefix+k, v)
+		if err != nil {
+			log.ErrorLog(ctx, "failed to set rbd qos, %s: %s. %v", k, v, err)
+
+			return err
+		}
+	}
+
+	return nil
+}
+
+func calcQosBasedOnCapacity(
```

Comment:
Can this also be a method of `rbdVol` or `rbdImage`?

Reply:
updated